httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java

```diff
@@ -0,0 +1,680 @@
+     * or exceptionally with an exception if all attempts fail
+     */
+    private CompletableFuture<ManagedAsyncClientConnection> connectAttempt(
+            final ConnectionInitiator connectionInitiator,
+            final HttpHost host,
+            final Timeout connectTimeout,
+            final Object attachment,
+            final List<InetAddress> addresses,
+            final SocketAddress localAddress) {
+
+        final CompletableFuture<ManagedAsyncClientConnection> connectionFuture = new CompletableFuture<>();
+
+        // Create a list of connection attempts to execute
+        final List<CompletableFuture<Void>> attempts = new ArrayList<>();
+        for (int i = 0; i < addresses.size(); i++) {
+            final InetAddress address = addresses.get(i);
+
+            if (LOG.isDebugEnabled()) {
+                LOG.info("Attempting to connect to {}", address);
+            }
+
+            final CompletableFuture<Void> attempt = new CompletableFuture<>();
+            attempts.add(attempt);
+            final HttpHost currentHost = new HttpHost(host.getSchemeName(), address, host.getHostName(), host.getPort());
+
+            connectionOperator.connect(
+                    connectionInitiator,
+                    currentHost,
+                    localAddress,
+                    connectTimeout,
+                    attachment,
+                    new FutureCallback<ManagedAsyncClientConnection>() {
+                        @Override
+                        public void completed(final ManagedAsyncClientConnection connection) {
+                            if (LOG.isDebugEnabled()) {
+                                LOG.debug("Successfully connected {}", ConnPoolSupport.getId(connection));
+                            }
+                            connectionFuture.complete(connection);
+                        }
+
+                        @Override
+                        public void failed(final Exception ex) {
+                            if (LOG.isDebugEnabled()) {
+                                LOG.debug("Failed to connect  {}", ConnPoolSupport.getId(address), ex);
+                            }
+                            attempt.completeExceptionally(ex);
+                        }
+
+                        @Override
+                        public void cancelled() {
+                            if (LOG.isDebugEnabled()) {
+                                LOG.debug("Cancelled connect for {}", ConnPoolSupport.getId(address));
+                            }
+                            attempt.cancel(true);
+                        }
+                    });
+
+            // Introduce a delay before executing the next connection attempt
+            if (i < addresses.size() - 1) {
+                try {
+                    final Duration delay = calculateDelay(i);
+                    delay.wait();
```

Comment:
What thread is this code running in? This will block for up to 2,000 milliseconds.

Reply:
The RFC recommends introducing a delay before starting the next connection attempt, but it doesn't mandate a specific implementation. The RFC suggests a default delay of 250 milliseconds, but also notes that a more nuanced implementation can use historical RTT data to influence the delay. Therefore, it is up to the implementation to decide whether to introduce a delay and how to calculate that delay.